File: page-objects/order-confirmation-page.js
const { By, until } = require('selenium-webdriver');

class OrderConfirmationPage {
  constructor(driver) {
    this.driver = driver;
  }

  // Locators
  getThankYouMessage() {
    return By.xpath(
      "//h1[contains(text(), 'Thank You')] | //h2[contains(text(), 'Thank You')] | //span[contains(text(), 'Thank You')] | " +
      "//h1[contains(text(), 'Order')] | //div[contains(text(), 'Order Confirmed')]"
    );
  }

  getOrderNumber() {
    return By.xpath(
      "//span[contains(text(), 'Order')] | //div[contains(@class, 'order-number')] | //p[contains(text(), '#')]"
    );
  }

  getConfirmationDetails() {
    return By.xpath(
      "//div[contains(@class, 'confirmation')] | //section[contains(@class, 'confirmation')] | " +
      "//div[contains(@class, 'order-details')]"
    );
  }

  getContinueShoppingButton() {
    return By.xpath(
      "//button[contains(text(), 'Continue')] | //a[contains(text(), 'Shop')] | //button[contains(text(), 'Back')]"
    );
  }

  getOrderStatus() {
    return By.xpath("//span[contains(@class, 'status')] | //div[contains(@class, 'status')]");
  }

  getEmailConfirmation() {
    return By.xpath("//span[contains(text(), 'email')] | //p[contains(text(), 'email')] | //text()[contains(translate(., 'CONFIRMATION EMAIL', 'confirmation email'), 'confirmation email')]/..");
  }

  getFirstNameText() {
    return By.xpath(
      "//span[contains(text(), 'John')] | //p[contains(text(), 'John')] | " +
      "//div[contains(text(), 'first') or contains(text(), 'First')]"
    );
  }

  getLastNameText() {
    return By.xpath(
      "//span[contains(text(), 'Doe')] | //p[contains(text(), 'Doe')] | " +
      "//div[contains(text(), 'last') or contains(text(), 'Last')]"
    );
  }

  getConfirmationEmailMessage() {
    return By.xpath(
      "//text()[contains(translate(., 'CONFIRMATION EMAIL SOON', 'confirmation email soon'), 'confirmation email soon')]/.. | " +
      "//p[contains(translate(., 'CONFIRMATION', 'confirmation'), 'confirmation') and contains(translate(., 'EMAIL', 'email'), 'email')] | " +
      "//div[contains(translate(., 'CONFIRMATION', 'confirmation'), 'confirmation')]"
    );
  }

  // Actions
  async waitForConfirmationPageLoad() {
    const thankYouMsg = await this.driver.wait(
      until.elementLocated(this.getThankYouMessage()),
      15000
    );
    await this.driver.wait(until.elementIsVisible(thankYouMsg), 10000);
  }

  async getThankYouText() {
    const message = await this.driver.wait(
      until.elementLocated(this.getThankYouMessage()),
      10000
    );
    return await message.getText();
  }

  async isThankYouMessageVisible() {
    try {
      const message = await this.driver.findElement(this.getThankYouMessage());
      return await message.isDisplayed();
    } catch (err) {
      return false;
    }
  }

  async getOrderDetails() {
    try {
      const details = await this.driver.findElement(this.getConfirmationDetails());
      return await details.getText();
    } catch (err) {
      return 'Order details not found';
    }
  }

  async extractOrderNumber() {
    try {
      await this.driver.wait(until.elementLocated(this.getOrderNumber()), 10000);
      const pageText = await this.driver.executeScript('return document.body.innerText');
      
      // Try to extract order number from the page text
      // Look for patterns like "Order #123456" or "Order Number: 123456"
      const orderPatterns = [
        /Order\s*#?(\d+)/i,
        /Order\s*Number[:\s]+(\d+)/i,
        /Ref\s*#?(\d+)/i,
        /#(\d{6,})/
      ];
      
      for (const pattern of orderPatterns) {
        const match = pageText.match(pattern);
        if (match && match[1]) {
          console.log(`Extracted Order Number: ${match[1]}`);
          return match[1];
        }
      }
      
      // Fallback to element text
      const orderElement = await this.driver.findElement(this.getOrderNumber());
      const orderText = await orderElement.getText();
      console.log(`Order Element Text: ${orderText}`);
      return orderText;
    } catch (err) {
      console.warn(`Unable to extract order number: ${err.message}`);
      return 'Order number not found';
    }
  }

  async isOrderConfirmed() {
    try {
      const thankYouMsg = await this.driver.findElement(this.getThankYouMessage());
      const isVisible = await thankYouMsg.isDisplayed();
      const text = await thankYouMsg.getText();
      return isVisible && (text.toLowerCase().includes('thank you') || text.toLowerCase().includes('confirmed'));
    } catch (err) {
      return false;
    }
  }

  async clickContinueShopping() {
    try {
      const continueBtn = await this.driver.findElement(this.getContinueShoppingButton());
      if (await continueBtn.isDisplayed()) {
        await continueBtn.click();
      }
    } catch (err) {
      // Button might not exist
    }
  }

  async validateCustomerNameOnConfirmation(firstName, lastName) {
    try {
      const pageText = await this.driver.executeScript('return document.body.innerText');
      const hasFirstName = pageText.includes(firstName);
      const hasLastName = pageText.includes(lastName);
      
      console.log(`First Name '${firstName}' visible: ${hasFirstName}`);
      console.log(`Last Name '${lastName}' visible: ${hasLastName}`);
      
      return hasFirstName && hasLastName;
    } catch (err) {
      console.warn(`Error validating customer names: ${err.message}`);
      return false;
    }
  }

  async validateEmailConfirmationMessage() {
    try {
      const pageText = await this.driver.executeScript('return document.body.innerText');
      const hasEmailMessage = pageText.toLowerCase().includes('confirmation email') || 
                              pageText.toLowerCase().includes("you'll receive");
      
      console.log(`Email confirmation message visible: ${hasEmailMessage}`);
      if (hasEmailMessage) {
        const emailMatch = pageText.match(/you['']ll\s+receive[^.!?]*email[^.!?]*[.!?]/i);
        if (emailMatch) {
          console.log(`Email message: ${emailMatch[0]}`);
        }
      }
      
      return hasEmailMessage;
    } catch (err) {
      console.warn(`Error validating email confirmation message: ${err.message}`);
      return false;
    }
  }

  async getStoredOrderNumber() {
    // This retrieves a previously extracted order number
    try {
      return global.storedOrderNumber || 'Order number not stored';
    } catch (err) {
      return 'Order number not found';
    }
  }
}

module.exports = OrderConfirmationPage;

File: page-objects/OrderConfirmationPage.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;

import java.time.Duration;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class OrderConfirmationPage {

    /** Set by the "extract and store the order number" step, replacing global.storedOrderNumber. */
    public static String storedOrderNumber;

    private final WebDriver driver;

    public OrderConfirmationPage(WebDriver driver) {
        this.driver = driver;
    }

    // Locators
    public By getThankYouMessage() {
        return By.xpath("//h1[contains(text(), 'Thank You')] | //h2[contains(text(), 'Thank You')]"
                + " | //span[contains(text(), 'Thank You')]"
                + " | //h1[contains(text(), 'Order')] | //div[contains(text(), 'Order Confirmed')]");
    }

    public By getOrderNumber() {
        return By.xpath("//span[contains(text(), 'Order')] | //div[contains(@class, 'order-number')]"
                + " | //p[contains(text(), '#')]");
    }

    public By getConfirmationDetails() {
        return By.xpath("//div[contains(@class, 'confirmation')] | //section[contains(@class, 'confirmation')]"
                + " | //div[contains(@class, 'order-details')]");
    }

    public By getContinueShoppingButton() {
        return By.xpath("//button[contains(text(), 'Continue')] | //a[contains(text(), 'Shop')]"
                + " | //button[contains(text(), 'Back')]");
    }

    public By getOrderStatus() {
        return By.xpath("//span[contains(@class, 'status')] | //div[contains(@class, 'status')]");
    }

    public By getEmailConfirmation() {
        return By.xpath("//span[contains(text(), 'email')] | //p[contains(text(), 'email')]"
                + " | //text()[contains(translate(., 'CONFIRMATION EMAIL', 'confirmation email'), 'confirmation email')]/..");
    }

    public By getFirstNameText() {
        return By.xpath("//span[contains(text(), 'John')] | //p[contains(text(), 'John')]"
                + " | //div[contains(text(), 'first') or contains(text(), 'First')]");
    }

    public By getLastNameText() {
        return By.xpath("//span[contains(text(), 'Doe')] | //p[contains(text(), 'Doe')]"
                + " | //div[contains(text(), 'last') or contains(text(), 'Last')]");
    }

    public By getConfirmationEmailMessage() {
        return By.xpath("//text()[contains(translate(., 'CONFIRMATION EMAIL SOON', 'confirmation email soon'), "
                + "'confirmation email soon')]/.."
                + " | //p[contains(translate(., 'CONFIRMATION', 'confirmation'), 'confirmation') "
                + "and contains(translate(., 'EMAIL', 'email'), 'email')]"
                + " | //div[contains(translate(., 'CONFIRMATION', 'confirmation'), 'confirmation')]");
    }

    // Actions
    public void waitForConfirmationPageLoad() {
        Helpers.wait(driver, Duration.ofSeconds(15))
                .until(ExpectedConditions.visibilityOfElementLocated(getThankYouMessage()));
    }

    public String getThankYouText() {
        WebElement message = Helpers.wait(driver, Duration.ofSeconds(10))
                .until(ExpectedConditions.presenceOfElementLocated(getThankYouMessage()));
        return message.getText();
    }

    public boolean isThankYouMessageVisible() {
        return Helpers.isVisible(driver, getThankYouMessage());
    }

    public String getOrderDetails() {
        try {
            return driver.findElement(getConfirmationDetails()).getText();
        } catch (RuntimeException err) {
            return "Order details not found";
        }
    }

    public String extractOrderNumber() {
        try {
            Helpers.wait(driver, Duration.ofSeconds(10))
                    .until(ExpectedConditions.presenceOfElementLocated(getOrderNumber()));
            String pageText = Helpers.bodyText(driver);

            // Try to extract order number from the page text
            // Look for patterns like "Order #123456" or "Order Number: 123456"
            List<Pattern> orderPatterns = Arrays.asList(
                    Pattern.compile("Order\\s*#?(\\d+)", Pattern.CASE_INSENSITIVE),
                    Pattern.compile("Order\\s*Number[:\\s]+(\\d+)", Pattern.CASE_INSENSITIVE),
                    Pattern.compile("Ref\\s*#?(\\d+)", Pattern.CASE_INSENSITIVE),
                    Pattern.compile("#(\\d{6,})"));

            for (Pattern pattern : orderPatterns) {
                Matcher match = pattern.matcher(pageText);
                if (match.find() && match.group(1) != null) {
                    System.out.println("Extracted Order Number: " + match.group(1));
                    return match.group(1);
                }
            }

            // Fallback to element text
            String orderText = driver.findElement(getOrderNumber()).getText();
            System.out.println("Order Element Text: " + orderText);
            return orderText;
        } catch (RuntimeException err) {
            System.out.println("Unable to extract order number: " + err.getMessage());
            return "Order number not found";
        }
    }

    public boolean isOrderConfirmed() {
        try {
            WebElement thankYouMsg = driver.findElement(getThankYouMessage());
            boolean isVisible = thankYouMsg.isDisplayed();
            String text = thankYouMsg.getText().toLowerCase();
            return isVisible && (text.contains("thank you") || text.contains("confirmed"));
        } catch (RuntimeException err) {
            return false;
        }
    }

    public void clickContinueShopping() {
        try {
            WebElement continueBtn = driver.findElement(getContinueShoppingButton());
            if (continueBtn.isDisplayed()) {
                continueBtn.click();
            }
        } catch (RuntimeException err) {
            // Button might not exist
        }
    }

    public boolean validateCustomerNameOnConfirmation(String firstName, String lastName) {
        try {
            String pageText = Helpers.bodyText(driver);
            boolean hasFirstName = pageText.contains(firstName);
            boolean hasLastName = pageText.contains(lastName);

            System.out.println("First Name '" + firstName + "' visible: " + hasFirstName);
            System.out.println("Last Name '" + lastName + "' visible: " + hasLastName);

            return hasFirstName && hasLastName;
        } catch (RuntimeException err) {
            System.out.println("Error validating customer names: " + err.getMessage());
            return false;
        }
    }

    public boolean validateEmailConfirmationMessage() {
        try {
            String pageText = Helpers.bodyText(driver);
            String lower = pageText.toLowerCase();
            boolean hasEmailMessage = lower.contains("confirmation email") || lower.contains("you'll receive");

            System.out.println("Email confirmation message visible: " + hasEmailMessage);
            if (hasEmailMessage) {
                Matcher emailMatch = Pattern
                        .compile("you['’]ll\\s+receive[^.!?]*email[^.!?]*[.!?]", Pattern.CASE_INSENSITIVE)
                        .matcher(pageText);
                if (emailMatch.find()) {
                    System.out.println("Email message: " + emailMatch.group());
                }
            }

            return hasEmailMessage;
        } catch (RuntimeException err) {
            System.out.println("Error validating email confirmation message: " + err.getMessage());
            return false;
        }
    }

    public String getStoredOrderNumber() {
        return storedOrderNumber == null ? "Order number not stored" : storedOrderNumber;
    }
}
